kernel/src/hil/time.rs

```diff
@@ -93,8 +113,33 @@ pub trait Client {
 /// The `Timer` trait models a timer that can notify when a particular interval
 /// has elapsed.
 pub trait Timer: Time {
+    fn set_client(&self, client: &'static TimerClient);
+
     /// Sets a one-shot timer to fire in `interval` clock-tics.
     fn oneshot(&self, interval: u32);
     /// Sets repeating timer to fire every `interval` clock-tics.
     fn repeat(&self, interval: u32);
+
+    // Q(alevy): Implementing this might require an additional, unnecessary field if a repeating
+    // timer is distinguished by having a non-zero value in the reload register. What if the return
+    // value is `Option<u32>` and `None` means it's oneshot, `Some(interval)` means it's repeating.
+    // Side benefit, `is_oneshot` and `is_repeating` can have default implementations.
+    fn interval(&self) -> u32;
+
+    fn is_oneshot(&self) -> bool;
+    fn is_repeating(&self) -> bool;
+
+    // This should return an option. Again, `is_enabled` can have a default implementation
+    fn time_remaining(&self) -> u32; // Returns 0 if disabled
+
+    fn is_enabled(&self) -> bool;
+
+    // Q(alevy): what are possible return values? why would you not be able to cancel a timer?
+    fn cancel(&self) -> ReturnCode;
```

Comment:
I think you can't cancel a timer if it's a one-shot that already fired -- there's a small race condition where the alarm triggers between the request to cancel starting and actually getting to the meat of cancellation

The simpler case is if someone calls cancel and there's no active timer (a bug, sure, but also nothing to cancel)

Reply:
(both of these scenarios would probably return `EALREADY`)

Reply:
The ReturnCode of cancel indicates whether you should expect a callback or not. A cancel can FAIL if the callback is already pending. Otherwise, you have to support logic to cancel outstanding callbacks, which is much more complex than the caller just handing the case that the callback will occur.

Reply:
Bubbling up the updated comment from Amit here since the github UI doesn't show the update by default:

```rust
    /// Cancels an outstanding timer.
    ///
    /// The implementation will _always_ cancel the timer, however, it may be possible for a timer
    /// to have already expired but not delivered to the client. In this case, the implementation
    /// must return [`FAIL`] letting the caller know that an event for the timer will still be
    /// delivered.

    // Q:(alevy) would it be simpler semantics to require the implementation to handle this?
    // Specifically, if there is a pending interrupt when `cancel` is called, the implementation
    // should ensure that a callback for the timer event is _not_ delivered. On the SAM4L, at
    // least, that would be nearly as easy to implement and seems easier to use.
```

I'm concerned that there are hard corner cases that make having the `FAIL` option appealing. I'm thinking of situations where the timer fires just as the userspace->kernel transition for the cancel syscall occurs, so the kernel hasn't started cancelling yet, nor handling the syscall, but the userspace has finished "making" the syscall, so now the timer frame gets pushed by the kernel (because we `service_pending_interrupts` before iterating processes) and then we start handling the process events, which have the cancel syscall, which would then have to _unstack_ the timer event.

I might have that exact sequence of events wrong, but I think this is one of those things where it can be a surprisingly onerous _guarantee_ for the kernel to have to provide, and having the out may prove useful, if possibly not used in the immediate architecture.